Update Security Guidelines Section (#27)
wso2/docs-security: en/docs/security-guidelines/secure-engineering-guidelines/dynamic-analysis-with-owasp-zap.md

## OWASP ZAP Setup

### Increase JVM Heap Size for Running ZAP
The heap size is defined in the *zap.sh* (for Linux) and *zap.bat* (for Windows) files. The default value is set to `Xmx512m` (if available free memory is above 1,500 MB) and increase the value appropriately based on the memory availability of your system. (At least 4GB is recommended)

In addition to that, the ZAP scan is a long-running process. Therefore, it is better to run ZAP in a cloud instance or in a dedicated server to avoid any interruptions. 

### Fine Tune ZAP Tool with Pre-Configured Policy 

> The ZAP tool should be fine-tuned before running a scan for obtaining better results. For this, you can download the WSO2 policy file for ZAP[^2] which contains the settings to fine-tune ZAP. 

Reviewer: ```suggestion
The ZAP tool should be fine-tuned before running a scan for obtaining better results. For this, you can download the WSO2 policy file for ZAP[^2], which contains the settings to fine-tune ZAP. 
```

Go to **Analyze** &rarr; **Scan Policy** Manager in ZAP.

![Placeholder](../../assets/images/secure-coding-guidelines/zap-01.png)

In the **Scan Policy Manager** window, click **Import**.

![Placeholder](../../assets/images/secure-coding-guidelines/zap-02.png)

Browse and select the WSO2 Policy file you downloaded.

![Placeholder](../../assets/images/secure-coding-guidelines/zap-03.png)

Since the policy file is imported correctly, you can use this later when you run the spider and the scan. 

### Configuring ZAP Proxy to Trace Browser Traffic

> Rather than providing the URL of the WSO2 server and attacking the URL with ZAP, it is much more effective if we record the UI actions we do on the WSO2 server and let the ZAP tool capture the traffic which can then be used for performing attacks. 

Reviewer: ```suggestion
Rather than providing the URL of the WSO2 server and attacking the URL with ZAP, it is much more effective if we record the UI actions we do on the WSO2 server and let the ZAP tool capture the traffic that can then be used for performing attacks. 
```

Go to **Tools** &rarr; **Options** &rarr; **Local Proxy** and set the hostname/IP address and the port number for the proxy. *(In this example, the port is set to 7777 which is selected randomly)*

![Placeholder](../../assets/images/secure-coding-guidelines/zap-04.png)

> Now the ZAP tool is ready to capture the traffic going through the above set port number. The next step is to configure the browser to send traffic through this port number so that the ZAP tool can trace them.

Reviewer: ```suggestion
Now the ZAP tool is ready to capture the traffic going through the port (set above). The next step is to configure the browser to send traffic through this port number so that the ZAP tool can trace them.
```

In Firefox, go to **Edit** &rarr; **Preferences** and in the **Advanced** options, click **Settings** under the Network tab. 

Select the **Manual proxy configuration** and set the hostname/IP and the port number.

![Placeholder](../../assets/images/secure-coding-guidelines/zap-05.png)

If any exception for **localhost, 127.0.0.1** or the hostname of the WSO2 server you are trying to scan is given in **No Proxy for:** text box, remove them so ZAP can detect the traffic flow of that as well. 

Now, go to firefox and access the WSO2 server. You will see the SSL warning below because the traffic goes through the ZAP proxy. Add an exception to the site in the browser.

![Placeholder](../../assets/images/secure-coding-guidelines/zap-06.png)

Now the WSO2 Server URL should be opened in the browser. Also, it should be listed in ZAP under the sites. 

![Placeholder](../../assets/images/secure-coding-guidelines/zap-07.png)

> Select the Mode of the scan as **Protected Mode**. With this, you can choose which sites to be used for attacking. *(For example, if you are trying out the Federated Authentication scenario with WSO2 Identity Server and Facebook, under the Sites in ZAP tool, the Facebook website also will be listed to be attacked. In the scope of scanning, external websites should be removed. With the Protected Mode, you have the capability to select the Sites that should be attacked by the ZAP tool)*

Reviewer: ```suggestion
Select the Mode of the scan as **Protected Mode**. With this, you can choose which sites to be used for attacking. *(For example, if you are trying out the Federated Authentication scenario with WSO2 Identity Server and Facebook, under the Sites in ZAP tool, the Facebook website also will be listed to be attacked. In the scope of scanning, external websites should be removed. With the Protected Mode, you able to select the sites that should be attacked by the ZAP tool)*
```

## Scanning Process

### Excluding the Server Logout from Spider

> When we run the ZAP scan with an active user session, if ZAP executes the action to logout from the server in the middle of the scan, the actions that should be performed with a logged in user session would not be performed after that (because the active session is removed with logout action ). In order to avoid that, we need to exclude the logout action from the Spider.

Reviewer: ```suggestion
When we run the ZAP scan with an active user session, if ZAP executes the action to logout from the server in the middle of the scan, the actions that should be performed with a logged in user session would not be performed after that (because the active session is removed with the logout action ). In order to avoid that, we need to exclude the logout action from the Spider.
```

> For that, first, log in to the WSO2 Server and then log out so that the logout action is traced by ZAP. Then, find the **GET:logout_action.jsp** and right-click it and exclude it from Spider.

Reviewer: ```suggestion
For that, first, log in to the WSO2 server and then log out so that the logout action is traced by ZAP. Then, find the **GET:logout_action.jsp** and right-click it and exclude it from Spider.
```

![Placeholder](../../assets/images/secure-coding-guidelines/zap-10.png)

On the URLs ZAP discovered, it can perform attacks to find possible issues. We can improve the coverage of the scan by manually performing all actions in the browser (i.e in Management Console) so that ZAP discovers each flow that it can try attacking. 

You can also automate this by having selenium scripts for all UI actions. After setting the ZAP to act as the proxy, you can run selenium scripts so that in the browser, it automatically plays the actions you would do manually. For every action, ZAP will discover the URLs. 

### Removing Unnecessary Sites from Scan
When the ZAP is acting as the proxy, all the URLs that the browser calls will be traced under the Sites in ZAP. We need to remove external websites and select only the WSO2 Server’s scope for the scan. 

Right-click the **Site** that should be included in the scan and select **Include in Context** &rarr; **New Context** .

![Placeholder](../../assets/images/secure-coding-guidelines/zap-11.png)

> Then it will show the **Session Properties** window with the regex for including the URL patterns in the scan. Provide a name for the context so that we can identify the site uniquely when we have multiple sites added as different contexts. (For example, when you are testing a product like WSO2 API Manager, the Management Console, API Store and API Publisher can be three major sites where you can add each as a different context)

Reviewer: ```suggestion
Then it will show the **Session Properties** window with the regex for including the URL patterns in the scan. Provide a name for the context so that we can identify the site uniquely when we have multiple sites added as different contexts. For example, when you are testing a product like WSO2 API Manager, the Management Console, API Store, and API Publisher can be three major sites where you can add each as a different context.
```

When searching also you can search URLs that belong only to the context with the **Search all URLs** option enabled. 

![Placeholder](../../assets/images/secure-coding-guidelines/zap-16.png)

You can also filter the alerts that belong only to the contexts you have added with the **Show all URLs** option enabled.

![Placeholder](../../assets/images/secure-coding-guidelines/zap-17.png)

### Globally Excluding URL Patterns
When the ZAP tool starts crawling the site, it will increase the network traffic heavily. We can reduce the traffic by excluding the URL patterns globally so that ZAP will ignore such URLs when crawling. For example, if we exclude URLs of .mp4 video files, the ZAP tool will not download mp4 video files which saves network bandwidth.

Go to **Tools** &rarr; **Options** and select the Global Exclude URL option. By default, there are some patterns already added to ZAP. You can select all of them for exclusion. Additionally, if there is any URL pattern you need to exclude, you can add the regex for the URL as a new entry. 

![Placeholder](../../assets/images/secure-coding-guidelines/zap-18.png)

### Creating the Logged in User Session

> When running the spider to crawl the site, we have to let ZAP discover the URLs that are accessible only by logged-in users as well. For that, we need to create a logged-in user session in ZAP, so that same as a logged-in user can browse the URLs in a web browser, ZAP will be able to crawl through those URLs. 

Reviewer: ```suggestion
When running the spider to crawl the site, we have to let ZAP discover the URLs that are accessible only by logged-in users as well. For that, we need to create a logged-in user session in ZAP, so that in the same way that a logged-in user can browse the URLs in a web browser, ZAP will be able to crawl through those URLs. 
```

Click  **Show All Tabs** in the toolbox so that it will display all the tabs. 

![Placeholder](../../assets/images/secure-coding-guidelines/zap-19.png)

Go to the **Http Sessions** tab. If there are already created sessions listed, you can remove them. 

![Placeholder](../../assets/images/secure-coding-guidelines/zap-20.png)

> Now while the ZAP proxy is tracing the traffic, go to the browser and log in to the site you need to scan. (When ZAP performs the scan, it will attack the URLs with the associated privileges of the user you logged in to). Once you log in, the session ID should be listed in the **HTTP Sessions** tab. Right-click the session and **Set as Active**. 

Reviewer: ```suggestion
Now, while the ZAP proxy is tracing the traffic, go to the browser and log in to the site you need to scan. Note that when ZAP performs the scan, it will attack the URLs with the associated privileges of the user you logged in as. Once you log in, the session ID should be listed in the **HTTP Sessions** tab. Right-click the session and **Set as Active**. 
```
